src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1015,73 +1005,116 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
 
 
 /*
- * RouterModifyTask builds a Task to represent a modification performed by
- * the provided query against the provided shard interval. This task contains
- * shard-extended deparsed SQL to be run during execution.
+ * RouterModifyTasks builds a List of Tasks representing a modification per-
+ * formed by the provided query using the provided list of modify routes.
+ * If possible, all tasks contain shard-specific SQL to be sent to workers
+ * during execution.
  */
-static Task *
-RouterModifyTask(Oid distributedTableId, Query *originalQuery,
-				 ShardInterval *shardInterval)
+static List *
+RouterModifyTasks(Oid distributedTableId, Query *query, List *modifyRouteList)
 {
-	Task *modifyTask = NULL;
+	List *modifyTasks = NIL;
 	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	ListCell *modifyRouteCell = NULL;
 
-	modifyTask = CitusMakeNode(Task);
-	modifyTask->jobId = INVALID_JOB_ID;
-	modifyTask->taskId = INVALID_TASK_ID;
-	modifyTask->taskType = MODIFY_TASK;
-	modifyTask->queryString = NULL;
-	modifyTask->anchorShardId = INVALID_SHARD_ID;
-	modifyTask->dependedTaskList = NIL;
-	modifyTask->replicationModel = cacheEntry->replicationModel;
-
-	if (originalQuery->onConflict != NULL)
+	foreach(modifyRouteCell, modifyRouteList)
 	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		ModifyRoute *modifyRoute = (ModifyRoute *) lfirst(modifyRouteCell);
+		Task *modifyTask = CitusMakeNode(Task);
+
+		modifyTask->jobId = INVALID_JOB_ID;
+		modifyTask->taskId = INVALID_TASK_ID;
+		modifyTask->taskType = MODIFY_TASK;
+		modifyTask->queryString = NULL;
+		modifyTask->anchorShardId = INVALID_SHARD_ID;
+		modifyTask->dependedTaskList = NIL;
+		modifyTask->replicationModel = cacheEntry->replicationModel;
+		modifyTask->valuesRTEIndex = 0;
+		modifyTask->rowValuesLists = NIL;
+
+		if (query->onConflict != NULL)
+		{
+			RangeTblEntry *rangeTableEntry = NULL;
 
-		/* set the flag */
-		modifyTask->upsertQuery = true;
+			/* set the flag */
+			modifyTask->upsertQuery = true;
 
-		/* setting an alias simplifies deparsing of UPSERTs */
-		rangeTableEntry = linitial(originalQuery->rtable);
-		if (rangeTableEntry->alias == NULL)
+			/* setting an alias simplifies deparsing of UPSERTs */
+			rangeTableEntry = linitial(query->rtable);
+			if (rangeTableEntry->alias == NULL)
+			{
+				Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+				rangeTableEntry->alias = alias;
+			}
+		}
+
+		if (modifyRoute != NULL)
```

Comment:
Having a partially populated task that we complete in the executor was a bit of the hack to get function evaluation in the partition column. It probably makes more sense to leave the route / task list empty when shard pruning is not yet possible and rebuild the task list in the executor.

e.g.
https://github.com/citusdata/citus/pull/1474/files#diff-4c4f0976727d06684376ecb22e8f5362R1123
https://github.com/citusdata/citus/pull/1474/files#diff-94b1a72e077cfc917cafbafd2a8b80ccR409